src/System.Linq/src/System/Linq/Enumerable.cs

```diff
@@ -573,9 +573,19 @@ public TResult[] ToArray()
                 }
                 return results;
             }
+
+            public List<TResult> ToList()
+            {
+                var results = new List<TResult>(_source.Length);
+                for (int i = 0; i < _source.Length; i++)
+                {
```

Comment:
You need to capture _source into a local here to let JIT optimize the range check. Otherwise JIT must assume that _source may change on every iteration and will not optimize.